sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,672 @@
+  that expression is not in a statement that is discarded through
+  `constexpr if`.  In this case, the function *F* uses all of the aspects named
+  in that attribute.
+
+* The function *F* contains a potentially evaluated expression that does any of
+  the following with a type *T* that is decorated with the
+  `[[sycl_detail::uses_aspects()]]` attribute, and that expression is not in a
+  statement that is discarded through `constexpr if`.  In this case, the
+  function *F* uses all of the aspects named in that attribute:
+
+  - Defines an object (including a temporary object) of type *U*.
+  - Calls a "new expression" of type *U*.
+  - Throws an expression of type *U*.
+  - Contains a cast to type *U*.
+  - References a literal of type *T*.
+
+  Where the type *U* is any of the following:
+
+  - The type *T*.
+  - A cv-qualified version of type *T*.
+  - An array of, pointer to, or reference to type *T*.
+  - A type that derives from type *T*.
+  - A class type that contains a non-static member object of type *T*.
+  - Any type that applies these rules recursively to type *T* (e.g. array of
+    pointers to type *T*, etc.)
+
+  When applying these rules, the front-end treats any use of the `double` type
+  as though it was implicitly decorated with
+  `[[sycl_detail::uses_aspects(has(aspect::fp64))]]`.
+
+If the `[[sycl_detail::uses_aspects()]]` attribute decorates a base class
+member function and a derived class overrides the member function, the
+overriding member function does not automatically inherit the attribute.
+Therefore, when the front-end considers the set of aspects used by a call to a
+member function, it need not consider any `[[sycl_detail::uses_aspects()]]`
+attributes that decorate overridden versions of the function.
+
+As noted earlier, standard SYCL does not allow indirect function calls or
+virtual functions in device code, although a DPC++ extension that adds some
+limited form of indirect function call is being contemplated.  If this
+extension allows virtual functions, we expect that when
+`[[sycl_detail::uses_aspects()]]` decorates a virtual function, it applies only
+to the static type of the class.  Therefore, when the front-end considers the
+set of aspects used by a virtual function call like the following:
+
+```
+void foo(Base *b) {
+  b->bar();
+}
+```
+
+It considers only the `[[sycl_detail::uses_aspects()]]` attribute that may
+decorate the definition of `Base::foo()` even though the application may pass a
+pointer to a derived class which decorates `foo()` differently.
+
+
+### Changes to other phases of clang
+
+Any clang phases that do function inlining will need to be changed, so that the
+`!intel_allowed_aspects` and `!intel_uses_aspects` metadata are transferred
+from the inlined function to the function that receives the inlined function
+body.  Presumably, there is already similar logic for the existing
+`!reqd_work_group_size` metadata, which already decorates device functions.
```

Comment:
I don't know about OpenCL, but this attribute was also used in SYCL 1.2.1, where it could appear on any device function (not just the kernel).  I presume that clang could inline these device function into the kernel that calls them, right?

Reply:
I changed the design in da9d05e9a9462006e64953b476be663457b14dea such that the metadata only needs to be preserved on kernel functions and on exported device functions, both of which have external linkage.  Therefore, there is no longer any danger that the metadata will be lost due to inlining.

If you think this issue is resolved, please mark the thread as resolved.

Reply:
I'm still not entirely convinced that this is completely safe, but I don't have enough knowledge and/or time to obtain it, so let's leave it as-is for now - we can change the approach later if it turns out to be impractical